38.md

```diff
@@ -0,0 +1,147 @@
+
+NIP-38
+======
+
+Encrypted Group Chat
+--------------------
+
+`draft` `optional` `author:vishalxl` 
+
+This NIP defines new event kinds for encrypted group chat. 
+
+It works by creating a shared secret that is used to encrypt the group or channel messages. Only the participants know that secret, so only such members can read/write to this group chat.
+
+It reserves three event kinds (140-142) for immediate use and 7 event kinds (143-149) for future use.
+
+- `140 - create encrypted channel`
+- `141 - change channel metadata including participants`
+- `142 - send encrypted message`
+
+
+## Kind 140: Create Encrypted channel
+
+Create a Encrypted chat channel.
+
+In the channel creation `content` field, Client SHOULD include basic channel metadata (`name`, `about` and `picture`).
+
+```json
+{
+    "content": "{\"name\": \"Demo Channel\", \"about\": \"A test channel.\", \"picture\": \"https://placekitten.com/200/200\"}",
+    ...
+}
+```
+
+The members of the group ( people who are allowed to post there) should be added as p tags to this event. The creator should also add themselves as p tag, if they want to participate there too.
+
+Further, the creator of the channel should send a Direct Message to each of the other participants, which will have the channel's shared-secret private/public keys. The Direct message should be sent by a kind other than 4; but presently kind 4 direct messages are used. 
+
+The format of that message is:
```

Comment:
we really shouldn't do this over kind 4. this is really hacky. ideally it would be a protocol message that clients can handle without parsing, display an invite properly, etc. Users shouldn't have to see a shared secret at all.

Reply:
Agree with this that there is no good reason to use kind 4 here. 

Invite suggestions on what kind it should be. I think it would be good to have a special kind where different apps can store or send their information from where they can pick it up. So that user can know what app is sending what message ( given that its the user itself who is sending these messages, or the client is doing so on user's behalf) 

Also then the format of the message, or its 'header' or first few words etc, should help differentiate between different apps.

Reply:
Edit the proposed NIP 38 and mentioned kind 104 rather than kind 4 to communicate the shared-secret.

Reply:
There is not much parsing involved for shared-secret message. Once the app matches the first few words, which are "App Encrypted Channels: ", then the required information is at well-known points in the message. If strInvite is the given message as a string then the required values are always at the same place:

Encrypted Channel id : strInvite.substring(58, 58 + 64) 
Secret private key        : strInvite.substring(159, 159 + 64); 
Secret public key        : strInvite:substring(224, 224 + 64);

And the length of strInvite should be 288 chars.

Reply:
I now think the above message should ideally be in json format. will work on it further in that direction. 

Unless you have some other idea there?

Reply:
no need for public key, it can be derived

i'll throw up a nip for "more secret" dm's